README.md
Debug Hints Library
===================

Copyright (C) 2002-2022 [std._debug authors][authors]

[![License](http://img.shields.io/:license-mit-blue.svg)](http://mit-license.org)
[![workflow status](https://github.com/lua-stdlib/_debug/actions/workflows/spec.yml/badge.svg?branch=master)](https://github.com/lua-stdlib/_debug/actions)
[![codecov.io](https://codecov.io/gh/lua-stdlib/_debug/branch/master/graph/badge.svg)](https://codecov.io/gh/lua-stdlib/_debug)


This is a debug hints management library for Lua 5.1 (including LuaJIT),
5.2, 5.3 and 5.4. The library is copyright by its authors (see the [AUTHORS][]
file for details), and released under the [MIT license][mit] (the same
license as Lua itself). There is no warranty.

`std._debug` has no run-time prerequisites beyond a standard Lua system.

[authors]: http://github.com/lua-stdlib/_debug/blob/master/AUTHORS.md
[github]: http://github.com/lua-stdlib/_debug/ "Github repository"
[lua]: http://www.lua.org "The Lua Project"
[mit]: http://mit-license.org "MIT License"


Installation
------------

The simplest and best way to install this library is with [LuaRocks][].
To install the latest release (recommended):

```bash
    luarocks install std._debug
```

To install current git master (for testing, before submitting a bug
report for example):

```bash
    luarocks install http://raw.githubusercontent.com/lua-stdlib/_debug/master/stdlib-git-1.rockspec
```

The best way to install without [LuaRocks][] is to copy the `std`
folder and its contents into a directory on your package search path.

[luarocks]: http://www.luarocks.org "Lua package manager"


Documentation
-------------

The latest release of these libraries is [documented in LDoc][github.io].
Pre-built HTML files are included in the release.

[github.io]: http://lua-stdlib.github.io/_debug


Bug reports and code contributions
----------------------------------

These libraries are written and maintained by their users.

Please make bug reports and suggestions as [GitHub Issues][issues].
Pull requests are especially appreciated.

But first, please check that your issue has not already been reported by
someone else, and that it is not already fixed by [master][github] in
preparation for the next release (see Installation section above for how
to temporarily install master with [LuaRocks][]).

There is no strict coding style, but please bear in mind the following
points when proposing changes:

0. Follow existing code. There are a lot of useful patterns and avoided
   traps there.

1. 3-character indentation using SPACES in Lua sources: It makes rogue
   TABS easier to see, and lines up nicely with 'if' and 'end' keywords.

2. Simple strings are easiest to type using single-quote delimiters,
   saving double-quotes for where a string contains apostrophes.

3. Save horizontal space by only using SPACES where the parser requires
   them.

4. Use vertical space to separate out compound statements to help the
   coverage reports discover untested lines.

[issues]: http://github.com/lua-stdlib/_debug/issues


NEWS.md
# Stdlib NEWS - User visible changes

## Noteworthy changes in release ?.? (????-??-??) [?]

### Bug fixes

  - Correct self arguments in metamethod declarations.


## Noteworthy changes in release 1.0.1 (2018-09-22) [stable]

### New Features

  - Initial support for Lua 5.4.


## Noteworthy changes in release 1.0 (2017-10-14) [stable]

### New features (since lua-stdlib-41.2)

  - Initial release, now separated out from lua-stdlib.

  - An entirely new API, which does not use global `_DEBUG` at all.


spec/spec_helper.lua
--[[
 Debug Hints Library for Lua 5.1, 5.2, 5.3 & 5.4
 Copyright (C) 2014-2022 std._debug authors
]]
local typecheck
have_typecheck, typecheck = pcall(require, 'typecheck')

local inprocess = require 'specl.inprocess'
local hell = require 'specl.shell'

badargs = require 'specl.badargs'

local cwd = os.getenv 'PWD'

package.path = cwd .. '/lib/?.lua;' .. cwd .. '/lib/?/init.lua;' .. package.path


-- Allow user override of LUA binary used by hell.spawn, falling
-- back to environment PATH search for 'lua' if nothing else works.
local LUA = os.getenv 'LUA' or 'lua'


-- Allow use of bare 'pack' and 'unpack' even in Lua 5.3.
pack = table.pack or function(...) return {n = select('#', ...), ...} end
unpack = table.unpack or unpack


local function getmetamethod(x, n)
   local m =(getmetatable(x) or {})[tostring(n)]
   if type(m) == 'function' then
      return m
   end
   if type((getmetatable(m) or {}).__call) == 'function' then
      return m
   end
end


function callable(x)
   return type(x) == 'function' or getmetamethod(x, '__call')
end


function copy(t)
   local r = {}
   for k, v in next, t do r[k] = v end
   return r
end


-- In case we're not using a bleeding edge release of Specl...
_diagnose = badargs.diagnose
badargs.diagnose = function(...)
   if have_typecheck then
      return _diagnose(...)
   end
end


-- A copy of base.lua:type, so that an unloadable base.lua doesn't
-- prevent everything else from working.
function objtype(o)
   return(getmetatable(o) or {})._type or io.type(o) or type(o)
end


local function mkscript(code)
   local f = os.tmpname()
   local h = io.open(f, 'w')
   -- TODO: Move this into specl, or expose arguments so that we can
   --          turn this on and off based on specl `--coverage` arg.
   h:write "pcall(require, 'luacov')"
   h:write(code)
   h:close()
   return f
end


--- Run some Lua code with the given arguments and input.
-- @string code valid Lua code
-- @tparam[opt={}] string|table arg single argument, or table of
--    arguments for the script invocation.
-- @string[opt] stdin standard input contents for the script process
-- @treturn specl.shell.Process|nil status of resulting process if
--    execution was successful, otherwise nil
function luaproc(code, arg, stdin)
   local f = mkscript(code)
   if type(arg) ~= 'table' then arg = {arg} end
   local cmd = {LUA, f, unpack(arg)}
   -- inject env and stdin keys separately to avoid truncating `...` in
   -- cmd constructor
   cmd.env = { LUA_PATH=package.path, LUA_INIT='', LUA_INIT_5_2='' }
   cmd.stdin = stdin
   local proc = hell.spawn(cmd)
   os.remove(f)
   return proc
end


local function tabulate_output(code)
   local proc = luaproc(code)
   if proc.status ~= 0 then return error(proc.errout) end
   local r = {}
   proc.output:gsub('(%S*)[%s]*',
      function(x)
         if x ~= '' then r[x] = true end
      end)
   return r
end


--- Show changes to tables wrought by a require statement.
-- There are a few modes to this function, controlled by what named
-- arguments are given.   Lists new keys in T1 after `require "import"`:
--
--       show_apis {added_to=T1, by=import}
--
-- @tparam table argt one of the combinations above
-- @treturn table a list of keys according to criteria above
function show_apis(argt)
   return tabulate_output([[
      local before, after = {}, {}
      for k in pairs(]] .. argt.added_to .. [[) do
         before[k] = true
      end

      local M = require ']] .. argt.by .. [['
      for k in pairs(]] .. argt.added_to .. [[) do
         after[k] = true
      end

      for k in pairs(after) do
         if not before[k] then print(k) end
      end
   ]])
end


--[[ ================ ]]--
--[[ Tmpfile manager. ]]--
--[[ ================ ]]--


local tmpdir = string.gsub(
   os.getenv 'TMPDIR' or os.getenv 'TMP' or '/tmp',
   '([^/])/*$', '%1'
) .. '/normalize-' .. math.random(65536)


local function append(path, ...)
   local n = select('#', ...)
   if n > 0 then
      local fh = io.open(path, 'a')
      fh:write(table.concat({...}, '\n') .. '\n')
      fh:close()
   end
   return n
end


-- Create a temporary file.
-- @usage
--    h = Tmpfile()            -- empty generated filename
--    h = Tmpfile(content) -- save *content* to generated filename
--    h = Tmpfile(name, line, line, line) -- write *line*s to *name*
Tmpfile = setmetatable({}, {
   _type = 'Tmpfile',

   __call = function(self, path_or_content, ...)
      local new = {}
      if select('#', ...) == 0 then
         new.path = os.tmpname()
         append(new.path, path_or_content)
      else
         new.path = path_or_content
         append(new.path, ...)
      end
      return setmetatable(new, getmetatable(self))
   end,

   __index = {
      dirname = function(self)
         return self.path:gsub('/[^/]*$', '', 1)
      end,

      basename = function(self)
         return self.path:gsub('.*/', '')
      end,

      append = function(self, ...)
         return append(self.path, ...)
      end,

      remove = function(self)
         return os.remove(self.path)
      end,
   },
})


Tmpdir = setmetatable({}, {
   _type = 'Tmpdir',

   __call = function(self, dirname)
      local new = {
         path = dirname or tmpdir,
         children = {},
      }
      os.execute("mkdir '" .. new.path .. "'")
      return setmetatable(new, getmetatable(self))
   end,

   __index = {
      file = function(self, name, ...)
         local child = Tmpfile(self.path .. '/' .. name, ...)
         self.children[#self.children + 1] = child
         return child
      end,

      subdir = function(self, name)
         local child = Tmpdir(self.path .. '/' .. name)
         self.children[#self.children + 1] = child
         return child
      end,

      remove = function(self)
         for _, child in ipairs(self.children) do
	child:remove()
         end
         return os.remove(self.path)
      end,
   }
})



lib/std/_debug/init.lua
--[[
 Debug Hints Library for Lua 5.1, 5.2, 5.3 & 5.4
 Copyright (C) 2002-2022 std._debug authors
]]
--[[--
 Manage debug mode, and associated substate hints.

 Set or change all the debug substate hints by calling the returned
 module functable with no argument to reset to defaults; with `true`
 to set all substate hints into development mode, or `false` for
 production mode.

    local _debug = require 'std._debug'(false)

 Query substate hints by indexing the returned module functable keys:

    local isstrict = _debug.strict

 Beware that even though you can change std._debug state at any time,
 stdlib libraries in particular (but probably other clients too) are
 configured at load time according to the state at the time they are
 required - e.g. changing _debug.strict after require 'std' does not
 affect the strict environment already created for the 'std' module
 when it was previously loaded.

 @module std._debug
]]



local _DEBUG = 'default'


--- Nothing!
local spec = {
   argcheck  = { default=true,  safe=true,    fast=false},
   deprecate = { default=nil,   safe=true,    fast=false},
   level     = { default=1,     safe=1,       fast=math.huge},
   strict    = { default=true,  safe=true,    fast=false},
}


local metatable = {
   --- Metamethods
   -- @section metamethods

   --- Change the all builtin debug hints.
   -- @function __call
   -- @bool[opt] enable or disable all debugging substate hints
   -- @treturn Substates substates
   -- @usage
   --   -- Enable all debugging substates
   --   local _debug = require 'std._debug'(true)
   __call = function(self, x)
      if x == true then
         _DEBUG = 'safe'
      elseif x == false then
         _DEBUG = 'fast'
      elseif x == nil then
         _DEBUG = 'default'
      else
         error("bad argument #1 to '_debug' (boolean or nil expected, got " .. type(x) .. ')', 2)
      end
      return self
   end,

   --- Lazy loading of std._debug modules.
   -- Don't load everything on initial startup, wait until first attempt
   -- to access a submodule, and then load it on demand.
   -- @function __index
   -- @string name submodule name
   -- @treturn table|nil the submodule that was loaded to satisfy the missing
   --    `name`, otherwise `nil` if nothing was found
   -- @usage
   --   local version = require 'std._debug'.version
   __index = function(_, name)
      local v = spec[name]
      if type(v) == 'table' then
         return v[_DEBUG]
      elseif v ~= nil then
         return v
      end
      local ok, r = pcall(require, 'std._debug.' .. name)
      if ok then
         spec[name] = r
         return r
      end
   end,
}


return setmetatable({}, metatable)


--- Types
-- @section types

--- Builtin debug substate hints.
--
-- Use `__call` metamethod to set all substate hints at once.
-- Note that none of the debugging features required to implement these
-- hints are encoded here, this module is merely a central location to
-- record systemwide hint; other modules you load subsequently may or may
-- not choose to behave according to their content.
-- @table Substates
-- @bool[opt=true] argcheck `true` if runtime argument checking is desired
-- @bool[opt] deprecate `nil` if deprecated api warnings are desired;
--   `false` if deprecated apis without warnings are desired; `true` if
--   removal of deprecated apis is preferred
-- @int[opt=1] level debugging level
-- @bool[opt=true] strict `true` if strict enforcement of variable declaration
--   before use is desired
-- @usage
--    require 'std._debug'.argcheck = false

